facebook.js

```diff
@@ -1,33 +1,127 @@
+				var linkTo = adContent.find('.userContent').next().find('a').attr('href');
+				if(linkTo.includes("l.facebook.com/l.php?")) {
+					linkTo = getParameterByName('u',linkTo);
+				} else {
+					linkTo = decodeURIComponent(linkTo);
+				}
+
+				// Snapshots are sent to storage, and kept in user session history too.
+				var snapshot = {
+					entity: advertiserName,
+					entityID: parseInt(advertiserID),
+					timestamp_created: parseInt(adContent.closest('[data-timestamp]').attr('data-timestamp')),
+					// Divide by 1000 to match the above, which is in seconds, compatibility for PHP
+					timestamp_snapshot: parseInt((Date.now() / 1000).toFixed()),
+					top_level_post_id: parseInt(top_level_post_id),
+					// May need to go through Facebook gateway, to get REAL url?
+					linkTo: linkTo,
+					postText: adContent.find('.userContent').text(),
+					// the big/small text beneath thumbnail images
+					// captures the majority of cases, but not all
+					fbStory_headline: adContent.find('.mbs._6m6._2cnj._5s6c, ._275z._5s6c').text(), // selectors for image, video
+					fbStory_subtitle: adContent.find('._6m7._3bt9, ._5q4r').text(),
+					// Maybe we want to download and save these on our server?
+					// Images are easy to store... but what about fb-locked videos?
+					thumbnailMedia: thumbnailMedia,
+				}
+
+				// snapshot_blobs are for server storage only.
+				var snapshot_blobs = {
+					html: adContent.html()
+				};
+
+				// Image snapshot
+				if(false) { // Bypass html2canvas for now, it's a tad crap
+					html2canvas(adContent, {
+						onrendered: function(canvas) {
+							snapshot_blobs.imageRender = canvas.toDataURL("image/png")
+							saveSnapshot();
+						}
+					});
+				} else {
+					saveSnapshot();
+				}
+
+				function saveSnapshot() {
+					// console.log("Noted an ad by "+snapshot.entity,snapshot.top_level_post_id)
+					updatedSessionHistory.push({'snapshot':snapshot,'snapshot_blobs':snapshot_blobs});
+					if(index == thisBatchN-1) {
+						synchronise(updatedSessionHistory);
+					}
+				}
+			}
+		})
+
+		function synchronise(updatedSessionHistory) {
+			console.log("Sync'ing ads just found ("+updatedSessionHistory.length+") w/ ads this session's history ("+currentSessionHistory.length+").");
+			// See if there are adverts new to the `currentSessionHistory`, to be POST'd and SESSION'd
+			if(updatedSessionHistory.length != currentSessionHistory.length) {
+				var newAdverts = updatedSessionHistory.diff(currentSessionHistory)
+				newAdverts.map(function(ad, index) {
+					// Only save small text to user session
+					userStorage.add('advertHistory', ad.snapshot);
+					console.log("New ad [USER SYNC'D] Advertiser: "+ad.snapshot.entity+" - Advert ID: "+ad.snapshot.top_level_post_id, ad.snapshot);
+					// Save the whole shabang to server
+					var wholeShabang = Object.assign({}, ad.snapshot, ad.snapshot_blobs);
+					$.post("https://who-targets-me.herokuapp.com/analytics/", wholeShabang, function( data ) {
```

Comment:
New logic to DB posting, which we need to reconcile. Now, the whole advert record is sent to the server, every (new) time it is seen.

Idk if this is exactly how we wanna do this, or if we wanna have a separate table for advert snapshots, assign an ID and then have a user-advertID frequency table or something? (I quite like the basic, user-entity frequency function.)